File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        def permutations(path, i):
            # base case
            if i == len(path) - 1:
                ans.append(path[:])
                return
            # rec case
            for j in range(i , len(path)):
                path[i], path[j] = path[j], path[i]
                permutations(path, i + 1)
                path[i], path[j] = path[j], path[i]
        permutations(nums, 0)
        return ans

File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> resultList = new ArrayList<>();
        
        backtrack(resultList,new ArrayList<>(),nums);
        return resultList;
    }
    
    private void backtrack(List<List<Integer>> resultList, ArrayList<Integer> tempList,int[] nums) {
        //base case
        if(tempList.size() == nums.length) {
            resultList.add(new ArrayList<>(tempList));
            return;
        }
        for(int number: nums) {
            //skip if we get same element
            if(tempList.contains(number)) continue;
            
            tempList.add(number);
            
            backtrack(resultList,tempList,nums);
            
            tempList.remove(tempList.size() - 1);
        }
    }
}